tests/validation/tests/v3_api/test_cluster_templates.py

```diff
@@ -361,6 +361,159 @@ def test_cluster_template_add_readonly_member():
+        clusterConfig=cluster_config1,
+        clusterTemplateId=cluster_template.id)
+
+    cluster_name = random_test_name("test-auto")
+    cluster = create_node_cluster(
+        client, name=cluster_name,
+        clusterTemplateRevisionId=cluster_template_revision1.id)
+    check_cluster_version(cluster, k8sversionlist[0])
+
+    # Reset the enforcement flag to false
+    data_test = {
+        "name": "cluster-template-enforcement",
+        "value": "false"
+    }
+    requests.put(enforcement_settings_url, json=data_test,
+                 verify=False, headers=headers)
+
+    cluster_cleanup(client, cluster)
+    cluster_cleanup(client, rkecluster)
+
+
+def test_cluster_template_enforcement_on_stduser():
+
+    # As an admin turn ON enforcement and ensure that standandard users
+    # can create clusters only using rke template. Creating clusters using
+    # regular rke config should not be allowed
+
+    enforcement_settings_url = CATTLE_TEST_URL + \
+                      "/v3/settings/cluster-template-enforcement"
+    data_test = {
+        "name": "cluster-template-enforcement",
+        "value": "true"
+    }
+    headers = {"Content-Type": "application/json",
+               "Accept": "application/json",
+               "Authorization": "Bearer " + ADMIN_TOKEN}
+    response = requests.put(enforcement_settings_url, json=data_test,
+                            verify=False, headers=headers)
+    print(response.content)
+    standard_user_client = \
+        get_client_for_token(user_token["user_standard"]["token"])
+    k8sversionlist = get_k8s_versionlist()
+    cluster_config1 = get_cluster_config(k8sversionlist[0])
+    rke_config = getRKEConfig(k8sversionlist[0])
+
+    # Verify creating cluster using rke template succeeds
+
+    cluster_template = standard_user_client.create_cluster_template(
+        name=random_test_name("template"), description="test-template")
+
+    cluster_template_revision1 = standard_user_client.\
+        create_cluster_template_revision(
+            name=random_test_name("revision1"),
+            clusterConfig=cluster_config1,
+            clusterTemplateId=cluster_template.id)
+
+    cluster_name = random_test_name("test-auto")
+    cluster = create_node_cluster(
+        standard_user_client, name=cluster_name,
+        clusterTemplateRevisionId=cluster_template_revision1.id)
+    check_cluster_version(cluster, k8sversionlist[0])
+
+    # Verify creating cluster using rkeconfig fails. API returns error as:
+    # "MissingRequired : A clusterTemplateRevision to create a cluster"
+
+    cluster_name = random_test_name("test-auto-rkeconfig")
+    with pytest.raises(ApiError) as e:
+        create_node_cluster(standard_user_client, cluster_name,
+                            rancherKubernetesEngineConfig=rke_config)
+    print(e)
+    assert e.value.error.status == 422
+    assert e.value.error.code == "MissingRequired"
+
+    # Reset the enforcement flag to false
```

Comment:
This block of code should be executed even when failure are encountered during the execution of this test.

Reply:
Added try and finally block.